### ISSUE TITLE
[ArangoDB 3.2][rocksdb] data racing when using more processes accessing same database

### ISSUE THREAD

**Issue by AxeeL** [arangodb/arangodb#2903]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [ ] 2.8
- [ ] 3.0
- [ ] 3.1
- [x] 3.2
- [ ] self-compiled devel branch

Mode:
- [ ] Cluster
- [x] Single-Server

Storage-Engine:
- [ ] mmfiles
- [x] rocksdb

On this operating system:
- [ ] DCOS on
  - [ ] AWS
  - [ ] Azure
  - [ ] own infrastructure
- [ ] Linux 
  - [ ] Debian .deb
  - [ ] Ubuntu .deb
  - [ ] SUSE   .rpm
  - [ ] RedHat .rpm
  - [ ] Fedora .rpm
  - [ ] Gentoo
  - [x] docker - official docker library
  - [ ] other:
- [ ] Windows, version: 
- [ ] MacOS, version:

Versions:
- [x] arangojs@5.7.0
- [x] node v6.11.1

After the release of 3.2.0 I wanted to try out the new rocksdb feature because it would help my case with huge datasets and memory issues. For first try running 3.2.0 with our application I got a weird error `ArangoError: precondition failed`, trying the default storage engine it was fine.
Today I decided to give it another try and track down the issue, but this time I got additional error `ArangoError: conflict (while executing)`

After a little testing I found out it's because of data racing on different processes trying to update database.

The issue can be reproduced by:
```js
'use strict';

const { fork } = require('child_process');

if (process.argv[2]) {
    const arangojs = require('arangojs');

    const connection = arangojs({
        url: 'http://root@localhost:8529',
        databaseName: 'local'
    });

    const query = `
        UPSERT {url: @url} 
        INSERT {url: @url, host: @host, baseDomain: @baseDomain, status: @status, socialKind: @socialKind} 
        UPDATE {host: @host, baseDomain: @baseDomain, socialKind: @socialKind} IN Link
        RETURN NEW._id
    `;

    const urls = 'abcdefghijklmnopqrstuvwxyz0123456789'.split('');

    console.log('Child forked!');
    const task = () => {
        const id = Math.floor(Math.random() * urls.length);
        const bindVars = {
            url: urls[id],
            host: "subdomain.exmaple.com",
            baseDomain: "exmaple.com",
            status: null,
            socialKind: true
        }
        connection
            .query(query, bindVars)
            .catch((e) => {
                console.log('ERROR:', e.message);
            })
    };

    let counter = 0;
    const interval = setInterval(() => {
        if (++counter >= 1e5) {
            clearInterval(interval);
            console.log('DONE')
            return;
        }

        if (counter % 100 === 0) {
            console.log('Counter at', counter);
        }

        task()
    }, 50);
} else {
    fork(process.argv[1], ['forked']);
    fork(process.argv[1], ['forked']);
}

```

If I run this using the default storage engine I get no errors, I can even set setInterval to 5 ms, using rocksdb I get results like this
```
Child forked!
Child forked!
ERROR conflict (while executing)
ERROR conflict (while executing)
ERROR conflict (while executing)
Counter at 100
Counter at 100
ERROR conflict (while executing)
ERROR conflict (while executing)
ERROR conflict (while executing)
Counter at 200
Counter at 200
ERROR conflict (while executing)
Counter at 300
Counter at 300
ERROR conflict (while executing)
^C
```
Is there any chance to solve this without touching application code?

Regarding `ArangoError: precondition failed`, even turning off the other process, this issue stays, please not for that query I'm using [collection.update](https://github.com/arangodb/arangojs#collectionupdate) and when I'm changing it to native AQL I get `ArangoError: conflict (while executing)`

**Comment by COLDTURNIP** [comment]
I meet the same issue (`'code': 409, 'errorNum': 1200, 'errorMessage': 'conflict (while executing)'`) while performing `UPSERT` via `pyArango.database.Database.AQLQuery` with multiprocessing.

**Comment by voondo** [comment]
Same error here (`ArangoError: precondition failed`) occuring randomly with arangodb 3.3.2, RocksDB, arangojs 5.8.0 and no AQL being directly involved (only `myCollection.update(key, ...)`).
I guess I have to return to mmfiles for now...

**Comment by m00ki3** [comment]
Getting the same during an update via query on rocksdb, 3.2.10, using the nodejs 5.7.1 driver

**Comment by irriss** [comment]
What does "ArangoError: conflict (while executing)" mean exactly in this example? 

```
const query = `
        UPSERT {url: @url} 
        INSERT {url: @url, host: @host, baseDomain: @baseDomain, status: @status, socialKind: @socialKind} 
        UPDATE {host: @host, baseDomain: @baseDomain, socialKind: @socialKind} IN Link
        RETURN NEW._id
```

Does it mean that UPSERT is not atomic operation?

Does it mean that UPSERT checked whether need to do INSERT or UPDATE and was about to do it but someone else changed the record in between?

Is it only UPSERT causing this particular error? What other scenarios can cause it?

Please provide as much information as possible. I am considering to add OPTION { ignoreErrors: true } (do I have other options?) but need to understand exactly what it will lead to.

**Comment by vizo** [comment]
Same here `ArangoError: AQL: conflict (while executing)` doing UPSERT

**Comment by sethyuan** [comment]
It's been half a year now, any updates on this bug?

**Comment by ohardy** [comment]
Hi,

I have the same problem on last version (3.3.9) while doing concurrent update on the same document.

Any workaround or fix ?

**Comment by ZStelmo** [comment]
And on 3.3.10

**Comment by jsteemann** [comment]
The RocksDB storage engine in ArangoDB *intentionally* does not use any collection locks, in order to allow for higher parallel throughput.
That means two concurrent requests/transactions may update the same document at the very same time, which is a write-write conflict that the storage engine will not allow. Otherwise any transaction could suffer from anomalies, which is probably undesired too.
The proper way to handle such conflicts is to retry them from the application side, because only the application will know if an operation is retryable or if retrying may make things worse.
However, in the `devel` branch we have augmented most of the AQL data modification statements with an `exclusive` option that will make them lock out any other write operations on the collection while running. This is opt-in however, because in many cases that is the opposite of what is actually desired. But in devel it's up to the user to decide.

**Comment by irriss** [comment]
@jsteemann Do I understand correctly that write-write conflict happens when a transaction tries to acquire write lock on a document already locked by another transaction/request?

**Comment by jsteemann** [comment]
Yes. Because if they would be let through, then one transaction would just overwrite data of another, without the client application getting a chance to detect this.

**Comment by irriss** [comment]
@jsteemann I have a specific use case, appreciate if you could advise the best approach. 
Let say I have an array of sensors, they send some values which I store in a database. For each sensor I am only interested in the latest value so when I receive new value from a sensor I run UPSERT command

`UPSERT { _key = sensorId }
INSERT { _key = sensorId, value = value_from_sensor }
UPDATE { _key = sensorId, value = value_from_sensor }`

Sometimes values from the same sensor come quickly then it can cause write-write conflict. If I handle it in application and try again after some delay it can cause overwriting of newer value. 

The proper solution probably would be a queue etc but for my task it seems excessive. 

I am wondering whether there is a setting for ArangoDb to just ignore write-write conflict and overwrite the value anyway?

**Comment by jsteemann** [comment]
@irriss : there are two solutions for this, but they are both available in our current devel version and will be released with ArangoDB 3.4:

* use "overwrite" option in AQL: `INSERT { _key: "someKey", ... } INTO collection OPTIONS { overwrite: true }`: this will turn the insert operation into a "repsert", which will try to insert the value if it does not yet exist, but replace and overwrite it if it exists
* use "exclusive" option in AQL: `UPSERT { ... } INSERT ... UPDATE ... INTO collection OPTIONS { exclusive: true }`: this will use an exclusive lock on the collection and block other operations while the upsert is going, thus preventing any write-write conflicts from happening.

For ArangoDB versions before 3.4 the solution is to use insert in a try ... catch block in the application, and react as follows:
* if it works, all good and done
* if it fails with "unique constraint violation", use a replace or update on the _key
* if it fails with any other error, let it bubble up
This is a bit more inconvenient, that's also why we have worked on the above alternatives in 3.4.

**Comment by AmmarMohamed** [comment]
@jsteemann  if i use mmfile engine the peoblem will be solved or not ?

**Comment by Simran-B** [comment]
The MMFiles engine uses collection level locks (writes block reads):
https://docs.arangodb.com/3.4/Manual/Architecture/StorageEngines.html

Therefore, the behavior with MMFiles engine should be like RocksDB with the `exclusive` option enabled.

**Comment by x3cion** [comment]
@jsteemann This is still a problem for me, even with `await collection.save(message, {overwrite: true, silent: true});` (ArangoJS) in ArangoDB 3.4.1. Can I help in any way to debug this or is this a known bug? Or is it not considered a bug?

**Comment by LJNeon** [comment]
I'm still getting data racing (on 3.6 with rocksdb) with a simple update request, `collection.update("key", {data}, {keepNull: false})`. Although it's error code 412 not 409.

**Comment by alexander-belikov** [comment]
`OPTIONS { exclusive: true }` for `INSERT` and `UPSERT` works for me for arangodb 3.8.7 (in case someone comes across this issue in 2022)
